File: check_pdns/check_pdns.py
#!/usr/bin/env python
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# You have to add these two lines with visudo to get this plugin working:
# Defaults:nrpe    !requiretty
# nrpe  ALL=NOPASSWD: /usr/bin/pdns_control
# The plugin stores some data in /tmp/check_pdns, if you have run it as a user
# which is not the monitoring user you have to delete the file so the
# monitoring user can read/write the file.
import sys
import re
import os
import time
import subprocess
import argparse
import shelve

class PowerDNSStats:
    __warning = False
    __critical = False
    __store_path = '/tmp/check_pdns'
    __values = ['packetcache-hit', 'packetcache-miss', 'query-cache-hit', 'query-cache-miss', 'recursing-answers', 'recursing-questions', 'corrupt-packets', 'servfail-packets', 'timedout-packets', 'udp-queries', 'tcp-queries', 'qsize-q']
    __data = dict()

    def __init__(self, warning=False, critical=False, warningerr=False, criticalerr=False):
        self.__warning = float(warning)
        self.__critical = float(critical)
        self.__warningerr = float(warningerr)
        self.__criticalerr = float(criticalerr)

    def __is_first_run(self):
	if not os.path.isfile(self.__store_path):
            self.__getData()
            self.__setData()
            return True
        else:
            return False

    def __getData(self):
        self.__data = self.__call_pdns__()
        self.__data['timeLastCheck'] = time.time()
        return self.__data


    def __setData(self):
        pluginStore = shelve.open(self.__store_path)
        pluginStore['pluginData'] = self.__data
        pluginStore.close()
        return True


    def __getStoredData(self):
        pluginDataFile = shelve.open(self.__store_path)
        pluginData = pluginDataFile['pluginData']
        pluginDataFile.close()
        return pluginData


    def __calcDiff(self, new_values, old_values):
        diff = new_values
        timediff = int(new_values['timeLastCheck']) - int(old_values['timeLastCheck'])
        for key in new_values:
            try:
                diff_value = (float(new_values[key]) - float(old_values[key])) / timediff
                if diff_value >= 0 and diff != False:
                    diff[key] = diff_value
                else:
                    diff = False
            except ZeroDivisionError as e:
                diff = 'TooFast'
        return diff



    def run(self):
        if not self.__is_first_run():
            new_values = self.__getData()
            old_values = self.__getStoredData()
            self.__setData()
            diff = self.__calcDiff(new_values, old_values)
            return self.__exit__(diff)
        else:
            return self.__exit__(False)


    def __call_pdns__(self):
        pdns_output = subprocess.Popen('LANG=en_EN.utf8 /usr/bin/sudo /usr/bin/pdns_control show "*"', shell=True,
                stdout=subprocess.PIPE).communicate()[0]
        return self.__process_pdns_output__(pdns_output)


    def __process_pdns_output__(self, pdns_output):
        for key in self.__values:
            value = re.search(key+'=?(\d*)', pdns_output).group(1)
            self.__data[key] = value

        return self.__data


    def __exit__(self, diff=False):
        return_code = 0
        prefix = "OK: "
        if diff == False:
            return_code = 3
            prefix = "UNKNOWN: "
            output = prefix + 'Collecting data for first time run'
        elif diff == "TooFast":
            return_code = 3
            prefix = "UNKNOWN: "
            output = prefix + 'Check can not be executed twice a second'
        else:
            # check if thresholds are set
            if self.__warningerr != False and self.__criticalerr != False:
                # check for critical errors
                if diff['corrupt-packets'] >= self.__criticalerr or diff['servfail-packets'] >= self.__criticalerr or diff['timedout-packets'] >= self.__criticalerr:
                    return_code = 2
                    prefix = "CRITICAL: "
                # check for warning errors
                elif diff['corrupt-packets'] >= self.__warningerr or diff['servfail-packets'] >= self.__warningerr or diff['timedout-packets'] >= self.__warningerr:
                    return_code = 1
                    prefix = "WARNING: "
            if self.__warning != False and self.__critical != False:
                # check for critical query count
                if (diff['udp-queries'] + diff['tcp-queries']) >= self.__critical:
                    return_code = 2
                    prefix = "CRITICAL: "
                # check for warning query count
                elif (diff['udp-queries'] + diff['tcp-queries']) >= self.__warning and return_code == 0:
                    return_code = 1
                    prefix = "WARNING: "

            output = prefix + 'Queries: %.3f/s udp-queries, %.3f/s tcp-queries, Error rates: %.3f/s servfail-packets, %.3f/s corrupt-packets, %.3f/s timedout-packets, Cache: %.3f/s packetcache-hit, %.3f/s packetcache-miss, %.3f/s query-cache-hit, %.3f/s query-cache-miss, Recursor: %.3f/s recursing-questions, %.3f/s recursing-answers' % (diff['udp-queries'], diff['tcp-queries'], diff['servfail-packets'], diff['corrupt-packets'], diff['timedout-packets'], diff['packetcache-hit'], diff['packetcache-miss'], diff['query-cache-hit'], diff['query-cache-miss'], diff['recursing-questions'], diff['recursing-answers']) + ' | '
            for key in diff:
                if key != 'timeLastCheck':
                    output += key + '=' + '%.3f' % diff[key] + ' '

        print output
        return return_code


def main():
    parser = argparse.ArgumentParser(description = 'Nagios plugin for monitoring PowerDNS')
    thresholds = parser.add_argument_group('query thresholds', 'warning and critical thresholds for queries per second')
    thresholds.add_argument('-w', '--warning', required=False, default=False,
            type=float, help='treshold for queries per second')
    thresholds.add_argument('-c', '--critical', required=False, default=False,
            type=float, help='treshold for queries per second')
    err_thresholds = parser.add_argument_group('error thresholds', 'warning and critical thresholds for errors per second')
    err_thresholds.add_argument('-W', '--warningerr', required=False, default=False,
            type=float, help='treshold for some kind of errors per second')
    err_thresholds.add_argument('-C', '--criticalerr', required=False, default=False,
            type=float, help='treshold for some kind of errors per second')
    args = parser.parse_args()

    powerdns_stats = PowerDNSStats(args.warning, args.critical, args.warningerr, args.criticalerr)
    return powerdns_stats.run()


if __name__ == '__main__':
    sys.exit(main())

File: check_pdns/check_pdns.php
<?php
$vertlabel = "stats per second";

$opt[1] = '';
$def[1] = '';
$opt[1] .= " --imgformat=PNG --title=\" $hostname / " . $this->MACRO['DISP_SERVICEDESC'] . "\" --base=1024 --vertical-label=\"$vertlabel\" --slope-mode -l 0 ";
$def[1] .= "DEF:ds2=$RRDFILE[2]:$DS[2]:AVERAGE " ;
$def[1] .= "DEF:ds7=$RRDFILE[7]:$DS[7]:AVERAGE " ;
$def[1] .= "CDEF:udp-queries=ds2 ";
$def[1] .= "CDEF:tcp-queries=ds7 ";

$ds_name[1] = 'Queries';

$def[1] .= "LINE1:udp-queries" . "#00CC00" . "FF:\"$NAME[2]\t\t\" " ;
$def[1] .= "GPRINT:udp-queries:LAST:\"Cur\\:%8.2lf $UNIT[2]\" ";
$def[1] .= "GPRINT:udp-queries:AVERAGE:\"Avg\\:%8.2lf $UNIT[2]\" ";
$def[1] .= "GPRINT:udp-queries:MIN:\"Min\\:%8.2lf $UNIT[2]\" ";
$def[1] .= "GPRINT:udp-queries:MAX:\"Max\\:%8.2lf $UNIT[2]\\n\" ";

$def[1] .= "LINE1:tcp-queries" . "#0000FF" . "FF:\"$NAME[7]\t\t\" " ;
$def[1] .= "GPRINT:tcp-queries:LAST:\"Cur\\:%8.2lf $UNIT[7]\" ";
$def[1] .= "GPRINT:tcp-queries:AVERAGE:\"Avg\\:%8.2lf $UNIT[7]\" ";
$def[1] .= "GPRINT:tcp-queries:MIN:\"Min\\:%8.2lf $UNIT[7]\" ";
$def[1] .= "GPRINT:tcp-queries:MAX:\"Max\\:%8.2lf $UNIT[7]\\n\" ";


$opt[2] = '';
$def[2] = '';
$opt[2] .= " --imgformat=PNG --title=\" $hostname / " . $this->MACRO['DISP_SERVICEDESC'] . "\" --base=1024 --vertical-label=\"$vertlabel\" --slope-mode -l 0 ";
$def[2] .= "DEF:ds1=$RRDFILE[1]:$DS[1]:AVERAGE " ;
$def[2] .= "DEF:ds9=$RRDFILE[9]:$DS[9]:AVERAGE " ;
$def[2] .= "DEF:ds10=$RRDFILE[10]:$DS[10]:AVERAGE " ;
$def[2] .= "CDEF:servfail-packets=ds1 ";
$def[2] .= "CDEF:timedout-packets=ds9 ";
$def[2] .= "CDEF:corrupt-packets=ds10 ";

$ds_name[2] = 'Errors';
$def[2] .= "LINE1:servfail-packets" . "#FF0000" . "FF:\"$NAME[1]\t\" " ;
$def[2] .= "GPRINT:servfail-packets:LAST:\"Cur\\:%8.2lf $UNIT[1]\" ";
$def[2] .= "GPRINT:servfail-packets:AVERAGE:\"Avg\\:%8.2lf $UNIT[1]\" ";
$def[2] .= "GPRINT:servfail-packets:MIN:\"Min\\:%8.2lf $UNIT[1]\" ";
$def[2] .= "GPRINT:servfail-packets:MAX:\"Max\\:%8.2lf $UNIT[1]\\n\" ";

$def[2] .= "LINE1:corrupt-packets" . "#00FF00" . "FF:\"$NAME[9]\t\" " ;
$def[2] .= "GPRINT:corrupt-packets:LAST:\"Cur\\:%8.2lf $UNIT[9]\" ";
$def[2] .= "GPRINT:corrupt-packets:AVERAGE:\"Avg\\:%8.2lf $UNIT[9]\" ";
$def[2] .= "GPRINT:corrupt-packets:MIN:\"Min\\:%8.2lf $UNIT[9]\" ";
$def[2] .= "GPRINT:corrupt-packets:MAX:\"Max\\:%8.2lf $UNIT[9]\\n\" ";

$def[2] .= "LINE1:timedout-packets" . "#0000FF" . "FF:\"$NAME[10]\t\" " ;
$def[2] .= "GPRINT:timedout-packets:LAST:\"Cur\\:%8.2lf $UNIT[10]\" ";
$def[2] .= "GPRINT:timedout-packets:AVERAGE:\"Avg\\:%8.2lf $UNIT[10]\" ";
$def[2] .= "GPRINT:timedout-packets:MIN:\"Min\\:%8.2lf $UNIT[10]\" ";
$def[2] .= "GPRINT:timedout-packets:MAX:\"Max\\:%8.2lf $UNIT[10]\\n\" ";


$opt[3] = '';
$def[3] = '';
$opt[3] .= " --imgformat=PNG --title=\" $hostname / " . $this->MACRO['DISP_SERVICEDESC'] . "\" --base=1024 --vertical-label=\"$vertlabel\" --slope-mode -l 0 ";
$def[3] .= "DEF:ds3=$RRDFILE[3]:$DS[3]:AVERAGE " ;
$def[3] .= "DEF:ds8=$RRDFILE[8]:$DS[8]:AVERAGE " ;
$def[3] .= "DEF:ds5=$RRDFILE[5]:$DS[5]:AVERAGE " ;
$def[3] .= "DEF:ds11=$RRDFILE[11]:$DS[11]:AVERAGE " ;
$def[3] .= "CDEF:query-cache-hit=ds3 ";
$def[3] .= "CDEF:query-cache-miss=ds8 ";
$def[3] .= "CDEF:packetcache-hit=ds5 ";
$def[3] .= "CDEF:packetcache-miss=ds11 ";

$ds_name[3] = 'Cache';
$def[3] .= "LINE1:query-cache-hit" . "#006600" . "FF:\"$NAME[3]\t\" " ;
$def[3] .= "GPRINT:query-cache-hit:LAST:\"Cur\\:%8.2lf $UNIT[3]\" ";
$def[3] .= "GPRINT:query-cache-hit:AVERAGE:\"Avg\\:%8.2lf $UNIT[3]\" ";
$def[3] .= "GPRINT:query-cache-hit:MIN:\"Min\\:%8.2lf $UNIT[3]\" ";
$def[3] .= "GPRINT:query-cache-hit:MAX:\"Max\\:%8.2lf $UNIT[3]\\n\" ";

$def[3] .= "LINE1:query-cache-miss" . "#00FF00" . "FF:\"$NAME[8]\t\" " ;
$def[3] .= "GPRINT:query-cache-miss:LAST:\"Cur\\:%8.2lf $UNIT[8]\" ";
$def[3] .= "GPRINT:query-cache-miss:AVERAGE:\"Avg\\:%8.2lf $UNIT[8]\" ";
$def[3] .= "GPRINT:query-cache-miss:MIN:\"Min\\:%8.2lf $UNIT[8]\" ";
$def[3] .= "GPRINT:query-cache-miss:MAX:\"Max\\:%8.2lf $UNIT[8]\\n\" ";

$def[3] .= "LINE1:packetcache-hit" . "#000066" . "FF:\"$NAME[5]\t\" " ;
$def[3] .= "GPRINT:packetcache-hit:LAST:\"Cur\\:%8.2lf $UNIT[5]\" ";
$def[3] .= "GPRINT:packetcache-hit:AVERAGE:\"Avg\\:%8.2lf $UNIT[5]\" ";
$def[3] .= "GPRINT:packetcache-hit:MIN:\"Min\\:%8.2lf $UNIT[5]\" ";
$def[3] .= "GPRINT:packetcache-hit:MAX:\"Max\\:%8.2lf $UNIT[5]\\n\" ";

$def[3] .= "LINE1:packetcache-miss" . "#9999FF" . "FF:\"$NAME[11]\t\" " ;
$def[3] .= "GPRINT:packetcache-miss:LAST:\"Cur\\:%8.2lf $UNIT[11]\" ";
$def[3] .= "GPRINT:packetcache-miss:AVERAGE:\"Avg\\:%8.2lf $UNIT[11]\" ";
$def[3] .= "GPRINT:packetcache-miss:MIN:\"Min\\:%8.2lf $UNIT[11]\" ";
$def[3] .= "GPRINT:packetcache-miss:MAX:\"Max\\:%8.2lf $UNIT[11]\\n\" ";



$opt[4] = '';
$def[4] = '';
$def[4] .= "DEF:ds4=$RRDFILE[4]:$DS[4]:AVERAGE " ;
$def[4] .= "DEF:ds6=$RRDFILE[6]:$DS[6]:AVERAGE " ;
$def[4] .= "CDEF:recursing-answers=ds4 ";
$def[4] .= "CDEF:recursing-questions=ds6 ";

$ds_name[4] = 'Recursor';
$def[4] .= "LINE1:recursing-questions" . "#660000" . "FF:\"".substr_replace($NAME[6], '', 17)."\t\" " ;
$def[4] .= "GPRINT:recursing-questions:LAST:\"Cur\\:%8.2lf $UNIT[6]\" ";
$def[4] .= "GPRINT:recursing-questions:AVERAGE:\"Avg\\:%8.2lf $UNIT[6]\" ";
$def[4] .= "GPRINT:recursing-questions:MIN:\"Min\\:%8.2lf $UNIT[6]\" ";
$def[4] .= "GPRINT:recursing-questions:MAX:\"Max\\:%8.2lf $UNIT[6]\\n\" ";

$def[4] .= "LINE1:recursing-answers" . "#FF0000" . "FF:\"$NAME[4]\t\" " ;
$def[4] .= "GPRINT:recursing-answers:LAST:\"Cur\\:%8.2lf $UNIT[4]\" ";
$def[4] .= "GPRINT:recursing-answers:AVERAGE:\"Avg\\:%8.2lf $UNIT[4]\" ";
$def[4] .= "GPRINT:recursing-answers:MIN:\"Min\\:%8.2lf $UNIT[4]\" ";
$def[4] .= "GPRINT:recursing-answers:MAX:\"Max\\:%8.2lf $UNIT[4]\\n\" ";

?>

